libtiledbsoma/src/utils/arrow_adapter.h

```diff
@@ -426,84 +426,186 @@ class ArrowAdapter {
         return get_table_string_column_by_index(arrow_table, index);
     }
 
+    /**
+     * Returns a copy of the data in a specified non-string column of an
+     * ArrowTable as a standard/non-Arrow C++ object.
+     */
     template <typename T>
-    static std::vector<T> get_table_column_by_index(
+    static std::vector<T> get_table_non_string_column_by_index(
         const ArrowTable& arrow_table, int64_t column_index) {
         ArrowArray* arrow_array = arrow_table.first.get();
         ArrowSchema* arrow_schema = arrow_table.second.get();
         _check_shapes(arrow_array, arrow_schema);
 
         if (std::is_same_v<T, std::string>) {
             throw std::runtime_error(
-                "SOMAArray::_core_domain_slot: template-specialization "
+                "SOMAArray::get_table_non_string_column_by_index: "
+                "template-specialization failure.");
+        }
+
+        ArrowArray* child_array = _get_and_check_column(
+            arrow_table, column_index, 2);
+
+        return get_array_non_string_column<T>(child_array);
+    }
```

Comment:
Could overload `get_table_non_string_column` but also fine as is.

```
template <typename T>
static std::vector<T> get_table_non_string_column(
    const ArrowTable& arrow_table, const std::string& column_name);

template <typename T>
static std::vector<T> get_table_non_string_column(
    const ArrowTable& arrow_table, int64_t column_index);
```

Reply:
> Could overload get_table_non_string_column but also fine as is.

I should leave a comment

Originally I used template specialization
```
template <typename T>
foo(T a, T b) 

template <>
foo(std::string a, std::string b)
```

and I kept getting spots where the non-string version was getting called even in string contexts

I tried to be clever & I persisted, but, ultimately I decided it wasn't worth it. I've been doing C/C++ a long time, I'm no slouch, and if it was fragile for me, it's not worth leaving in a fragile state.

So I went with the `_string` suffix, and the
```
if (std::is_same_v<T, std::string>) {
    throw std::runtime_error(...)
}
```

Reply:
@nguyenv https://github.com/single-cell-data/TileDB-SOMA/pull/3069/commits/e05b1ea60fbacd6c6860c340cd927f71eaea7b19